### ISSUE TITLE
[Bug Report] Multiply and divide has different broadcasting rules

### ISSUE THREAD

**Issue by rfurko-tt** [tenstorrent/tt-metal#12798]
**Describe the bug**
I can't use `ttnn.divide` same way as `ttnn.multiply`. Multiply works as expected, divide crashes. 

**To Reproduce**
```
import ttnn
import torch
import numpy as np

with ttnn.manage_device(device_id=0) as device:
    x = torch.ones((1, 2, 3, 4), dtype=torch.bfloat16)
    y = torch.ones((1, 1, 1, 1), dtype=torch.bfloat16) * 0.5
    
    x_tt = ttnn.from_torch(x, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
    y_tt = ttnn.from_torch(y, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)

    x_y_mult_tt = ttnn.multiply(x_tt, y_tt)
    print(ttnn.to_torch(x_y_mult_tt))

    x_y_divide_tt = ttnn.divide(x_tt, y_tt)
    print(ttnn.to_torch(x_y_divide_tt))

```

**Expected behavior**
Divide should have same broadcasting rules to multiply. 

**Please complete the following environment information:**
- OS: Ubuntu 20.04
- Version of software (eg. commit): 12fdb68d525ebc73c5e9e19db9a953e3640c5359

**Comment by dmakoviichuk-tt** [comment]
@eyonland I think you are owner of the unary/binary ops. Feel free to move to the right person.

**Comment by dmakoviichuk-tt** [comment]
HI @yugaoTT I've noticed that DIV is implemented as RECIP and MUL ops inside.
Looks like the case where broadcast should work was accidentally missed and seems easy to support it as everything is ready.

**Comment by eyonland** [comment]
@umadevimcw,  it looks like your solution would generalize for both inputs automatically and possibly solve this as a general case on eltwise ops.   I think the return on this is high and we quickly get to better coverage of this functionality.

**Comment by dmakoviichuk-tt** [comment]
@umadevimcw  changes are not acceptable. It should be solved in the normal way.
Yo rub one simple DIV op we added a few more ops.
@eyonland It is a really bad solution for a couple of reasons:
1) we are adding specific changes to the generic binary function
2) we are making division slower. If we want to make it slower we can always call reciprocal and than multiplication like what we are doing right now as a workaround.

**Comment by rfurko-tt** [comment]
@umadevimcw Thanks for providing detailed explanation. 
Could you please provide more information:
1. why `ttnn::divide` is executed as two operations? We are often work in `bfloat16` which suffers from lack of precision and I would expect every operation takes a little bit of precision from the result. 
2. why don't we re-use multiply broadcasting mechanism inside DIV_FAST? (In proposed solution we use redundant multiply to use it implicitly) 
3. I think it's fair to have bunch of workarounds in the model (temporarily), but we shouldn't have hacks on top of hacks in ops development. What's our current position on this? 
4. From performance perspective, when we perform divide we perform 3 operations: multiply, multiply, reciprocal (which is divide). I think it's unreasonable and be a hidden source of development slowdown. What's our policy here for adding overhead? 

Thanks in advance!

**Comment by eyonland** [comment]
@rfurko-tt and @dmakoviichuk-tt ,  if I understand this problem correctly, we cannot do broadcasting within the kernel operation itself because there is no clear way to identify the logical shape of the tiled tensor in the width and height dimensions.   We either need to allow this workaround to unblock models or we wait for the tensor layout and shape class to be rewritten and at that time we come back to this and fix the kernels to properly handle the broadcasting.    My position is that we allow this for now, knowing it is not an optimal solution, but then circle back and implement this properly in the kernel itself after the tensor layout is fixed.   @umadevimcw , please open a separate issue to track the work of implementing broadcasting in the kernel itself.  This needs to be done regardless of the outcome of this issue.

**Comment by dmakoviichuk-tt** [comment]
@eyonland 
You understood the problem incorrectly.
If broadcasting works in multiply it must work in div operation. There are no excuses.
In my opinion this hot fix doesn't make any sense. Because it makes it work even slower than obvious workaround with 2 manuall calls.
We should never allow "fixes" like this. In order to unblock the issue we can simply call reciprocal and multiply manually.
It is a bug/issue in the binary op implementation where in one case div is treated ad chain of RECIP and MUL but when it goes broadcast way there is no support of it in both TTNN and kernel side. I expected you as an owner to drive this issue because it will require proper fixes in TTNN binary op.

**Comment by mrakitaTT** [comment]
Hi folks, we are hitting this issue in tt-mlir compiler so I just wanted to check with you what is the status of this issue and do you maybe have some estimate how long it will take to fix it?

**Comment by eyonland** [comment]
The broadcasting generally is being fixed and will be delivered November 22, 2024.  Shortly afterwards we will look into how we support combining multiple operations for binary eltwise ops.  My expectation is currently the end of November for divide as it is a combination.

**Comment by prajaramanTT** [comment]
Keeping Priority label in sync with Priority field - P2

**Comment by mrakitaTT** [comment]
Hi @umadevimcw @eyonland I've noticed this has been moved to "In Progress", do you maybe have some estimate when you expect the fix to be in?

**Comment by umalesTT** [comment]
This is still an issue. Fails with ```BinaryOpType cannot be mapped to BcastOpMath```

**Comment by sdjordjevicTT** [comment]
@cmaryanTT can you please help us getting the status update of this. We are still experiencing failures with the DIV broadcast on our end.

**Comment by umadevimcw** [comment]
@sdjordjevicTT  Could you try using the ttnn.experimental.div operation instead of ttnn.div and let me know if you encounter any issues?

**Comment by sdjordjevicTT** [comment]
> [@sdjordjevicTT](https://github.com/sdjordjevicTT) Could you try using the `ttnn.experimental.div` and `mul` operation instead of `ttnn.div` and let me know if you encounter any issues?

Thank you for the suggestion, @umadevimcw! We can give that a try. Before we proceed, could you share the pros and cons of using experimental operations in general?

**Comment by cmaryanTT** [comment]
@sdjordjevicTT - in this case, we have a work-in-progress rewrite of the broadcast infrastructure in the experimental ops. The intent is that this will replace the non-experimental binary op in the near future. Experimental is generally some combination of OPs that are works-in-progress or are developed for specific usage and aren't suitable yet for general use.

**Comment by staylorTT** [comment]
@cmaryanTT do we have a timeline for the experimental ops to become the standard?

**Comment by cmaryanTT** [comment]
@staylorTT - for binary ops, we should switch it to default within a couple of weeks. Integration is currently under debug, with one outstanding major fix that needs to be implemented.

**Comment by sdjordjevicTT** [comment]
@cmaryanTT, any updates regarding the switch?

**Comment by cmaryanTT** [comment]
Keep using experimental. We had an attempt to turn it on this week, but had to revert to fix a perf issue.

**Comment by mrakitaTT** [comment]
@cmaryanTT Is there any way we can track the effort to turn it on again? Is there a GH issue for that perf degradation investigation?

**Comment by cmaryanTT** [comment]
We will update through @staylorTT when we have something committed and it's survived some live testing.

**Comment by staylorTT** [comment]
Can we please use this worklog to communicate the updates @cmaryanTT

**Comment by cmaryanTT** [comment]
This will be addressed by #16147 - follow that issue for updates.

**Comment by KalaivaniMCW** [comment]
Since this is scalar broadcast case, by explicitly enabling binary_ng with `use_legacy` flag, we can utilise the broadcast support for ops other than add/sub/mul. i.e.,  `x_y_divide_tt = ttnn.divide(x_tt, y_tt, use_legacy=False)`

<img width="879" alt="Image" src="https://github.com/user-attachments/assets/c59f4ccf-f1ab-4cfb-93cb-ec813c4f8c64" />

**Comment by cmaryanTT** [comment]
Looks like the same issue as the other case - scalar broadcast has issues with legacy binary. We should adjust the fallback criteria and add some missing tests. @dchenTT

**Comment by dchenTT** [comment]
Fixed by https://github.com/tenstorrent/tt-metal/pull/23686 No flag is required.

**Comment by skhatriTT** [comment]
Hey @dchenTT, could you please provide an update on this? Do we have an ETA?

**Comment by dchenTT** [comment]
It was fixed by this PR https://github.com/tenstorrent/tt-metal/pull/23686. However, it was reverted later by this PR. https://github.com/tenstorrent/tt-metal/pull/24561.  Right now there is no perfect  solution yet. The workaround is to use flag use_legacy=False.

**Comment by dchenTT** [comment]
If Forge cannot set the flag, then it has to wait for the reverted PR to be reverted back. However, it is blocked by this issue https://github.com/tenstorrent/tt-metal/issues/22308. Once #22308 is fixed soon, I will revert again.

**Comment by staylorTT** [comment]
https://github.com/tenstorrent/tt-metal/issues/22308 can a PR to fix it, we will wait for this to be stable in main before addressing this specific bug.

**Comment by dchenTT** [comment]
Please test with this PR (revert of revert) https://github.com/tenstorrent/tt-metal/pull/26101

**Comment by staylorTT** [comment]
@mrakitaTT to test and followup with findings. @umalesTT Please also look at this.

**Comment by umalesTT** [comment]
@staylorTT followup from Forge training side is that we no longer encounter errors related to this.

**Comment by fahmedchanTT** [comment]
@umalesTT Can this issue be closed, then? Is there other testing to be performed?

CC: @skhatriTT

**Comment by umalesTT** [comment]
@fahmedchanTT yes it can be closed.
